<a href="https://colab.research.google.com/github/adityabajoria/Complete-Python-3-Bootcamp/blob/master/Lecture_3_3_Pandas_Part_2_(Colab).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 3.3 Pandas Part 2
We will continue to learn how to use the Python data analysis library `pandas` to read and manipulate data.

# More pandas!



Let's use a weather dataset, which has a lot of numerical data. This will allow us to get numerical insights that were harder with the categorical data we saw in the Netflix dataset last lecture.

## Test your skills
* Read in the csv from the URL specified, and store it in a pandas dataframe called `df`.
* Display the dataframe (use the `display()` function)

In [ ]:
import pandas as pd

WEATHER_DATA_URL = 'https://raw.githubusercontent.com/alextsun/ssi-ds-bootcamp-2020/master/Week3/weatherHistory.csv'

df = pd.read_csv(WEATHER_DATA_URL)
display(df)

Formatted Date  ...                           Daily Summary
0      2006-04-01 00:00:00.000 +0200  ...       Partly cloudy throughout the day.
1      2006-04-01 01:00:00.000 +0200  ...       Partly cloudy throughout the day.
2      2006-04-01 02:00:00.000 +0200  ...       Partly cloudy throughout the day.
3      2006-04-01 03:00:00.000 +0200  ...       Partly cloudy throughout the day.
4      2006-04-01 04:00:00.000 +0200  ...       Partly cloudy throughout the day.
...                              ...  ...                                     ...
96448  2016-09-09 19:00:00.000 +0200  ...  Partly cloudy starting in the morning.
96449  2016-09-09 20:00:00.000 +0200  ...  Partly cloudy starting in the morning.
96450  2016-09-09 21:00:00.000 +0200  ...  Partly cloudy starting in the morning.
96451  2016-09-09 22:00:00.000 +0200  ...  Partly cloudy starting in the morning.
96452  2016-09-09 23:00:00.000 +0200  ...  Partly cloudy starting in the morning.

[96453 rows x 12 columns]

Notice that there are far more data points in this weather dataset than the Netflix dataset. For machine learning applications, you will likely have a lot of data (>100,000 data points) that you use to train your models.

Storing all these data points in arrays yourself is possible, but requires a lot more effort when you want to manipulate this data, and `pandas` abstracts this away and provides a lot of useful functions for data manipulation (as we began to see last lecture).


# More Indexing

Last time we practiced some indexing, but let's try some more indexing to get some better practice. You will find indexing to be one of the most useful aspects of working with dataframes, and you will likely use this for any research project involving data science and anaysis.

There are a few ways to index using pandas, and we saw a couple last time. Here, we will use a pandas attribute called `loc` that can be used on any dataframe.

You can use this attribute to look at only specific columns, as follows:

`df.loc[:,["Column A", "Column B"]]`

You can also filter your results. For example, if you want to see pressure for every day that both rained and had a temperature above 10 degrees Celcius, you could do the following:

`df.loc[(df["Precip Type"] == "rain") & (df["Temperature (C)"]>20.0), ["Precip Type", "Temperature (C)", "Pressure (millibars)"]]`

In the above code, the first argument is `(df["Precip Type"] == "rain") & (df["Temperature (C)"]>20.0)` and it describes the boolean conditions that need to be met for the row to be included in the returned dataframe.

The second argument is `["Precip Type", "Temperature (C)", "Pressure (millibars)"]` and it names the columns to be included in the returned dataframe (and the order of these columns)

Let's run it and see what it does!



In [ ]:
pressure_filtered = df.loc[(df["Precip Type"] == "rain") & (df["Temperature (C)"]>20.0), \
                           ["Precip Type", "Temperature (C)", "Pressure (millibars)"]]
display(pressure_filtered)

Precip Type  Temperature (C)  Pressure (millibars)
37           rain        20.044444               1012.22
38           rain        21.050000               1011.44
39           rain        21.183333               1010.52
40           rain        20.116667               1009.83
41           rain        20.216667               1009.26
...           ...              ...                   ...
96448        rain        26.016667               1014.36
96449        rain        24.583333               1015.16
96450        rain        22.038889               1015.66
96451        rain        21.522222               1015.95
96452        rain        20.438889               1016.16

[20391 rows x 3 columns]

You'll notice that, although there are only 20391 rows, the index label of the last row is 96452. This is because this index label corresponds to the its row in the original dataframe. Thus, you can use this index label to match rows from this filtered dataframe to the original weather dataframe.

## Test your skills
* Make a dataframe called `filtered_df` that includes all Partly Cloudy days where it rained and the humidity was greater than 0.8. Include the date, the summary, the precipitation type, the humidity, and the wind speed as the 5 columns in the dataframe (in that order).

In [ ]:
# TODO

Formatted Date  ... Wind Speed (km/h)
0      2006-04-01 00:00:00.000 +0200  ...           14.1197
1      2006-04-01 01:00:00.000 +0200  ...           14.2646
3      2006-04-01 03:00:00.000 +0200  ...           14.1036
5      2006-04-01 05:00:00.000 +0200  ...           13.9587
6      2006-04-01 06:00:00.000 +0200  ...           12.3648
...                              ...  ...               ...
96406  2016-09-08 01:00:00.000 +0200  ...            3.2200
96407  2016-09-08 02:00:00.000 +0200  ...            0.3703
96408  2016-09-08 03:00:00.000 +0200  ...            3.0268
96409  2016-09-08 04:00:00.000 +0200  ...            3.2039
96429  2016-09-09 00:00:00.000 +0200  ...            2.9624

[7891 rows x 5 columns]

# Getting statistical insights

As previously mentioned, one of the benefits of this weather dataset is there is a lot of numerical data. We discussed the benefits of numerical data previously.

Specifically, we learned about advantages of looking at mean and standard deviations of various data, and what they tell us about our data.

Pandas dataframes have a function called `describe()` that return a dataframe that contains a summary of these statistics for all the numerical data in our dataset.





## Test your skills
* Call `describe()` on the original weather dataframe and see the summary statistics that it outputs

In [ ]:
# TODO

Temperature (C)  ...  Pressure (millibars)
count     96453.000000  ...          96453.000000
mean         11.932678  ...           1003.235956
std           9.551546  ...            116.969906
min         -21.822222  ...              0.000000
25%           4.688889  ...           1011.900000
50%          12.000000  ...           1016.450000
75%          18.838889  ...           1021.090000
max          39.905556  ...           1046.380000

[8 rows x 8 columns]

## Getting central tendencies

We can find the mean of a particular column by getting a dataframe of that column, and using the `mean()` function.

You can get a dataframe of a column by using the indexing techniques we learned. Then you can call the `mean()` function, in which you specify which axis you would like to take the mean. Axis 0 means you take the mean along each column. Axis 1 means you take the mean across each row.

You can get the median and mode in the same way using the `median()` and `mode()` functions.

## Test your skills
* Find the mean across the `Temperature (C)` column. You should get a value of `11.93`.
* Similarly, find the median across the `Pressure (millibars)` column.
* Find the mode across the `Precip Type` column. Notice that although this isn't a numerical datatype itself, getting the mode of nonumerical data can still give useful information.


Hint: If you want to include all rows, your first argument (which are your boolean conditions) should be `:`. This specifies that all rows will meet the condition.

In [ ]:
# TODO

0    rain
dtype: object

## Getting measures of spread

We discussed how standard deviation tells you how much data varies, which can be a very important metric. This is one kind of measure of "spread".

We can find this for any column in a dataframe using the `std()` function (very similarly to how we found the mean, median, and mode previously).


## Test your skills
* Find the standard deviation of the `Wind Speed (km/h)` column

In [ ]:
# TODO

6.9135710125921515

# Grouping data

These measures of central tendency and spread are inherently limited in what they can tell us, because they operate across all data points.

For example, it seems intuitive that the average temperature may be strongly linked to the precipation type. For that reason, we may want average temperature of only when it rains and average temperature of only when it snows.

We can run a test to see whether these two things may actually be linked in the way we suspect.

You can do this using the `groupby()` function on the dataframe. As the argument to this function, pass in the the columns you want to group by in an array. This function doesn't return exactly a dataframe, but you can still use the central tendecy and spread measure functions on it. You do not need to specify an axis argument.

For example, below we get the `Summary` and `Humidity` columns using the `loc` method of indexing. Then we group by `Summary` and find the `mean()` across the humidities.

Set the `as_index` parameter to `True` in order to have the indices of each mean humidity be the associated summary.
Set the `as_index` parameter to `False` in order to have the indices of each summary and associated mean humidity to be numerically ordered, returning a dataframe with both columns accessible.


In [ ]:
summary_humidity_df = df.loc[:, ["Summary", "Humidity"]]
display(summary_humidity_df.groupby(["Summary"],as_index=False).mean())

Summary  Humidity
0                                Breezy  0.637778
1                        Breezy and Dry  0.260000
2                      Breezy and Foggy  0.938571
3              Breezy and Mostly Cloudy  0.637054
4                   Breezy and Overcast  0.763144
5              Breezy and Partly Cloudy  0.545803
6                                 Clear  0.729708
7   Dangerously Windy and Partly Cloudy  0.490000
8                               Drizzle  0.867949
9                                   Dry  0.230294
10                Dry and Mostly Cloudy  0.242143
11                Dry and Partly Cloudy  0.240814
12                                Foggy  0.950765
13              Humid and Mostly Cloudy  0.874250
14                   Humid and Overcast  0.881429
15              Humid and Partly Cloudy  0.848824
16                           Light Rain  0.888095
17                        Mostly Cloudy  0.725069
18                             Overcast  0.837232
19                        Partly Cloudy  0.648571
20                                 Rain  0.947000
21                                Windy  0.572500
22                        Windy and Dry  0.240000
23                      Windy and Foggy  0.900000
24              Windy and Mostly Cloudy  0.600000
25                   Windy and Overcast  0.708667
26              Windy and Partly Cloudy  0.528806

## Test your skills
* Run the same analysis as above, but group by `Precip Type` and get the mean `Temperature (C)`
* Then, get the std of `Temperature (C)`


In [ ]:
# TODO

Temperature (C)
Precip Type                 
rain               13.852989
snow               -3.270885

Temperature (C)
Precip Type                 
rain                8.317505
snow                3.075312

We can see from this that, as we originally suspected, temperature and precipation are very closely linked. The temperature is much higher when it is raining compared to when it is snowing.

This is the first step to identifying correlations, which is at the core of data science.


We can use this same kind of thinking to see if there are other relationships that we may not initially suspect. the `groupby()`, `mean()`, and `std()` functions are powerful and can work across entire dataframes (rather than specificying specific columns).

Let's give this a try.

## Test your skills
* Get the mean of all numerical columns grouped by `Precip Type`

Hint: You do not have to use `loc` to filter out specific columns. You can call `groupby()` and `mean()` on the entire dataframe.

In [ ]:
# TODO

Temperature (C)  ...  Pressure (millibars)
Precip Type                   ...                      
rain               13.852989  ...           1002.764548
snow               -3.270885  ...           1006.216970

[2 rows x 8 columns]

As you can see, grouping can be a very powerful tool in data analysis, and thus is great to have in your arsenal.

# Sorting data

Sometimes you may want to sort data as well. For example, maybe your data is a time series (meaning you are recording data over the course of some time period), and thus there is some natural ordering you would like to maintain and have represented in your data analysis.

Pandas has a lot of built-in sorting functionality for all your sorting needs!

Call `sort_values()` on your dataframe, and pass the columns you want to sort by. You can indicate a list of columns where the first column in the primary one to sort by, the second column is the secondary one to sort by, etc.

You pass in these columns you want to sort by as an argument called `by`. The default behavior is to sort these rows in ascending order. However, you can change the behavior to sort in descending order using the `ascending` argument. Set the `ascending` argument to a list equal to the length of the `by` argument, and for each column specific `True` for ascending order or `False` for descending order.

## Test your skills
* Sort the data by `Apparent Temperature (C)` primarily, and then by `Temperature (C)` secondarily.


In [ ]:
# TODO

Formatted Date  ...                                      Daily Summary
54864  2012-02-11 00:00:00.000 +0100  ...  Foggy starting overnight continuing until morn...
54845  2012-02-10 05:00:00.000 +0100  ...  Foggy starting in the morning continuing until...
54866  2012-02-11 02:00:00.000 +0100  ...  Foggy starting overnight continuing until morn...
55489  2012-02-09 01:00:00.000 +0100  ...                               Foggy until morning.
55492  2012-02-09 04:00:00.000 +0100  ...                               Foggy until morning.
...                              ...  ...                                                ...
65534  2013-07-29 14:00:00.000 +0200  ...                    Partly cloudy in the afternoon.
9903   2007-08-24 15:00:00.000 +0200  ...                       Partly cloudy until morning.
65535  2013-07-29 15:00:00.000 +0200  ...                    Partly cloudy in the afternoon.
65533  2013-07-29 13:00:00.000 +0200  ...                    Partly cloudy in the afternoon.
66111  2013-06-22 15:00:00.000 +0200  ...                         Partly cloudy until night.

[96453 rows x 12 columns]

# Joining/Merging Data (OPTIONAL)

Sometimes, your data may come from different sources, or you will want to add data to your dataset. Pandas has built-in merging and joining functionality that you can take advantage of to do this.

For example, let's say you want to create a new column that includes all of the temperatures at Fahrenheit. You can create this column and merge it with the existing data, using the `join()` function of a dataframe.

There are many ways that these join operations can happen, so I highly suggest reading the documentation/user guide when you are looking to perform your own merging operations (https://pandas.pydata.org/pandas-docs/stable/user_guide/merging.html). 

Here, we are using the most simple version, where the indices of the data we are joining match up. Thus, we can simply call `original_dataframe.join(new_dataframe)`, which returns the merged dataframe, in order to merge the data.

Let's look at how to do this with the Celsius to Fahrenheit conversion mentioned before.

In [ ]:
fahrenheit = (df.loc[:,["Temperature (C)"]] * 1.8) + 32
fahrenheit = fahrenheit.rename(columns={"Temperature (C)": "Temperature (F)"})

extended_df = df.join(fahrenheit)
display(extended_df)


Formatted Date  ... Temperature (F)
0      2006-04-01 00:00:00.000 +0200  ...           49.05
1      2006-04-01 01:00:00.000 +0200  ...           48.84
2      2006-04-01 02:00:00.000 +0200  ...           48.88
3      2006-04-01 03:00:00.000 +0200  ...           46.92
4      2006-04-01 04:00:00.000 +0200  ...           47.76
...                              ...  ...             ...
96448  2016-09-09 19:00:00.000 +0200  ...           78.83
96449  2016-09-09 20:00:00.000 +0200  ...           76.25
96450  2016-09-09 21:00:00.000 +0200  ...           71.67
96451  2016-09-09 22:00:00.000 +0200  ...           70.74
96452  2016-09-09 23:00:00.000 +0200  ...           68.79

[96453 rows x 13 columns]

Let's take a closer look at this code.

First, you create a dataframe with only the column you are looking to modify. You can see here that doing addition and multiplication with scalar values (single numbers) means that the operation is applied to every value in the dataframe. Thus, every Celsius temperature is converted to Fahrenheit.

We want to change the column name to reflect the unit change, so we use `dataframe.rename()` which takes in a mapping from old names to new names.

Then, finally, we run our join. The indices match already beacuse we used data with these indices to generate our new data. Thus, the join is very simple and clean.

## Test your skills
* Add a new column to the weather dataframe called `Wind Speed Normalized` that subtracts the mean wind speed from each data point and then divides each data point by the standard deviation of the wind speed.

In [ ]:
# TODO